### PR TITLE
announce indices for deals (rebased on main)

### PR DESCRIPTION
**Pull request by nonsense**
Should be able to announce indices and serve indices for Boost deals and legacy markets deals.
Markets PR at https://github.com/filecoin-project/go-fil-markets/pull/691.

### TODO

- [x] Boost should start after markets so the index provider callback isn't overwritten.
- [x] Test Market deals announcement on Sofia Miner.
- [x] Test Boost deal announcement on Sofia Miner.
- [x] Test the `AnnounceAllDeals` CLI on Sofia Miner.

### CODE REVIEW

File: cmd/boost/index.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+			Name:        "color",
+			Usage:       "use color in display output",
+			DefaultText: "depends on output being a TTY",
+		},
+	},
+	Action: func(cctx *cli.Context) error {
+		ctx := lcli.ReqContext(cctx)
+
+		// announce markets deals
+		nodeApi, closer, err := lcli.GetMarketsAPI(cctx)
```
Comment (nonsense) [inline cmd/boost/index.go:33]: @aarshkshah1992 I think this should be `GetBoostAPI`.

File: cmd/boost/index.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+		if err != nil {
+			return err
+		}
+		defer closer()
+		if err := nodeApi.IndexerAnnounceAllDeals(ctx); err != nil {
+			fmt.Printf("some errors while announcing legacy markets deals: %s", err)
+		}
```
Comment (nonsense) [inline cmd/boost/index.go:40]: @aarshkshah1992 this doesn't make sense to me. We will never have a `markets` and `boost` deployment at the same time. I guess we should remove this.

File: indexprovider/wrapper.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+		dealsDB:     dealsDB,
+		legacyProv:  legacyProv,
+		prov:        prov,
+		dagStore:    dagStore,
+		meshCreator: meshCreator,
+	}
+}
+
+func (w *Wrapper) IndexerAnnounceAllDeals(ctx context.Context) error {
+	log.Info("will announce all Boost deals to Indexer")
```
Comment (nonsense) [inline indexprovider/wrapper.go:44]: This is not very informative, maybe downgrade to `Debug` ?

File: cmd/boost/index.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+var indexProvAnnounceAllCmd = &cli.Command{
+	Name:  "announce-all",
+	Usage: "Announce all active deals to indexers so they can download the indices",
+	Flags: []cli.Flag{
+		&cli.BoolFlag{
+			Name:        "color",
+			Usage:       "use color in display output",
+			DefaultText: "depends on output being a TTY",
+		},
```
Comment (nonsense) [inline cmd/boost/index.go:25]:
```suggestion
```

File: indexprovider/wrapper.go [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+		legacyProv:  legacyProv,
+		prov:        prov,
+		dagStore:    dagStore,
+		meshCreator: meshCreator,
+	}
+}
+
+func (w *Wrapper) IndexerAnnounceAllDeals(ctx context.Context) error {
+	log.Info("will announce all Markets deals to Indexer")
+	err := w.legacyProv.AnnounceAllDealsToIndexer(ctx)
```
Comment (nonsense) [inline indexprovider/wrapper.go:45]: We don't seem to be doing anything with that error, maybe check it and have a WARN log message if it is != nil ?

File: cmd/boost/index.go (hunk 2 above)
Comment (aarshkshah1992) [reply cmd/boost/index.go]: This is done.

File: storagemarket/deal_execution.go [hunk 6]
```diff
@@ -498,11 +502,18 @@ func (p *Provider) indexAndAnnounce(ctx context.Context, pub event.Emitter, deal
...
+	if err := stores.RegisterShardSync(ctx, p.dagst, pc, deal.InboundFilePath, true); err != nil && !xerrors.Is(err, dagstore.ErrShardExists) {
 		return fmt.Errorf("failed to register deal with dagstore: %w", err)
 	}
 	p.dealLogger.Infow(deal.DealUuid, "deal successfully registered in dagstore")
 
+	// announce to the network indexer
+	annCid, err := p.ip.AnnounceBoostDeal(ctx, deal)
+	if err != nil && !xerrors.Is(err, provider.ErrAlreadyAdvertised) {
+		return fmt.Errorf("failed to announce deal to index provider: %w", err)
+	}
```
Comment (nonsense) [inline storagemarket/deal_execution.go:514]: Shouldn't we check for err != nil (i.e. when the err is provider.ErrAlreadyAdvertised), before saying that deal is successfully announced?

File: node/modules/storageminer.go [hunk 7]
```diff
@@ -463,39 +465,56 @@ func NewLogsDB(logsSqlDB *LogSqlDB) *db.LogsDB {
 	return db.NewLogsDB(logsSqlDB.db)
 }
 
+// We start Boost in this way because we want to ensure Boost is started after the legacy markets Provider.
+// We want to do this so that the index provider callback registered by Boost can overwrite the index
+// provider callback registered by legacy markets as legacy markets does NOT know about Boost storage deals.
```
Comment (dirkmc) [inline node/modules/storageminer.go:470]: Is this comment still correct? It looks like we're starting the index provider in `HandleIndexProvider`

**Review by dirkmc — APPROVED** [review]

File: indexprovider/wrapper.go (hunk 3 above)
Comment (aarshkshah1992) [reply indexprovider/wrapper.go:48]: It's just once per that CLI command. Harmless but nice to have when going through logs

File: storagemarket/deal_execution.go (hunk 6 above)
Comment (aarshkshah1992) [reply storagemarket/deal_execution.go:514]: Hey Anton,  if the error is `provider.ErrAlreadyAdvertised`, we have nothing to worry about and can consider announcement to be successful given how the index provider stuff works.

File: node/modules/storageminer.go (hunk 7 above)
Comment (aarshkshah1992) [reply node/modules/storageminer.go:470]: Removed.

Comment (aarshkshah1992) [reply node/modules/storageminer.go:470]: And added a comment to `HandleIndexProvider`.

File: storagemarket/deal_execution.go (hunk 6 above)
Comment (nonsense) [reply storagemarket/deal_execution.go:514]: Returning both a result and an error is certainly not idiomatic behaviour and somewhat confusing - this is why I mentioned the comment. We basically end up with an error and with the result (`annCid`), which is really not what one expects unless they know the full stack.

File: cmd/boost/index.go (hunk 4 above)
Comment (nonsense) [reply cmd/boost/index.go:25]: `color` is not used, so we can remove it.

File: storagemarket/deal_execution.go (hunk 6 above)
Comment (aarshkshah1992) [reply storagemarket/deal_execution.go:514]: I see what you mean. Have fixed this.

**Pull request merged**